### ISSUE TITLE
Creating Events

### ISSUE THREAD

**Issue by AleskiWeb** [Credera/angular-bing-maps#28]
Hi again,

I am trying to create an event for a 'viewchange' handler. However I can't seem to ever get into the function that is to be called when the handler detects a change. For example I am passing in the following object using the attribute for "data-events":
$scope.events = {
        viewChange: 'viewchange'
};

Inside the $scope.$watch('events', function (events) function inside the angular-bing-maps.js:
var bingMapsHandler = Microsoft.Maps.Events.addHandler($scope.map, eventName, function (event) {
    console.log(event);
    console.log('test');
    console.log('eventName');
    console.log('usersHandler);
    usersHandler(event);
    $scope.$apply();
});

None of the console logs are returning anything (not even undefinded) so I was wondering if there was anything amiss here? Would you have an example on the correct implementation as I am sure it's a problem on my end ofcourse :dancers:

**Comment by sjmorrow** [comment]
If I read your code correctly, I think it is because you are passing a string to the viewChange attribute on $scope.events. If you are looking to hook into the 'viewchange' event, your $scope.events would need to look like this:
```javascript
$scope.events = {
    viewchange: function(event) {
        console.log('Caught the viewchange event');
    }
}
```

Look at the following example for a working event handler 
https://github.com/Credera/angular-bing-maps/blob/master/example/infobox-directive.html#L25

As a side-note to the OP or anyone finding this in the future, if you are registering event listeners on a specific Entity (Like a pushpin), and need to determine which specific Entity the event was triggered on, we have added a 'track-by' directive attribute that allows you to pass in a unique identifier (which you will need to provide yourself). This unique identifier will be available on the event as follows:
```javascript
$scope.events = {
    click: function(event) {
        console.log('The unique identifier of the Entity clicked is: ' + event.target.trackBy);
    }
}
```
We are currently doing this in our Polygon directive example, so look there for a complete code sample.
https://github.com/Credera/angular-bing-maps/blob/master/example/polygon-directive.html

**Comment by AleskiWeb** [comment]
Seriously appreciate the communication you deliver on here. Thanks for being a STAR!

**Comment by sjmorrow** [comment]
Unfortunately no, not at this time. I did some digging and Bing Maps does not expose the zoomLevel and mapCenter via any of their events (and we just pass the native Bing Maps event through to your handlers). To get the zoomLevel and mapCenter attributes, you would need access to the Microsoft.Maps.Map() instance that we create and call the Map.getCenter() and Map.getZoom() methods. We are planning on making a new service for angular-bing-maps that would allow the user access to this very thing, as well as any other low-level Bing Maps functionality. I'll create a new issue for this functionality to make sure we get it in a future release.

**Comment by AleskiWeb** [comment]
What would the theory/plan be for this? I'd love to have a crack at doing it but I wouldn't really know where to start. Do you mean service in the typical sense (provider/service/factory)? Would this just be set when the map instantiates in the bingMap directive and then exposed to the application so you can reference it (like 'var map = ServiceThatStoresMap.map) so that you can then run a .getZoomLevel on that?

**Comment by sjmorrow** [comment]
I was just about to submit this comment saying you were correct when I realized that solution actually breaks down if the user adds multiple <bing-map> directives to a single page. How would we know which one to return from the service? After further investigation, we should create a new attribute / scope value on the <bing-map> directive that accepts and fires a callback once the map has been instantiated. This will allow the user to create a function in their controller that can grab the reference to the Microsoft.Maps.Map. 

Feel free to take a stab at it, and I can look at it later if you get stuck. Go to [this page](http://www.ng-newsletter.com/posts/directives.html) and search for "Parent execution binding" to see an example of what needs to be done. We will need to modify /src/angular-bing-maps/directives/map.directive.js to accept this new scope attribute, and execute the function once we have finished all of our setup.

I would recommend naming the new scope attribute 'onMapReady', which would equate to the user's directive looking like so:
```html
<bing-map on-map-ready="mapReadyCallback(map)">
```
